core/src/BoundaryCondition.cc

```diff
@@ -154,6 +196,145 @@ void VelocityBoundaryCondition::evaluateImpulseCondition(
+// Evaluate the new particle positions on the boundary.
+// (-x = 0, +x = 1, -y = 2, +y = 3, -z = 4, +z = 5)
+void PeriodicBoundaryCondition::evaluateBoundaryPosition(
+    const std::shared_ptr<Mesh>& mesh, const int boundary,
+    std::vector<Particle>& particles ) const
+{
+    // Determine if boundary is on x, y, or z.
+    int dim = std::floor( boundary / 2 );
+
+    // Get the length of the mesh in the current dimension.
+    double length = mesh->getDimensionLength( dim );
+
+    // If left boundary, move particles to right side.
+    if ( boundary % 2 == 0 )
+    {
+        for ( auto& p : particles )
+            if ( p.r[dim] < 0.0 )
+		p.r[dim] += length;
+    }
+    // If right boundary, move particles to left side. 
+    else
+    {
+	for ( auto& p : particles )
+	    if ( p.r[dim] > length )
+	        p.r[dim] -= length;
+    }	
+}
+
+//---------------------------------------------------------------------------//
+// Complete Boundary sum
+// (-x = 0, +x = 1, -y = 2, +y = 3, -z = 4, +z = 5)
+void PeriodicBoundaryCondition::completeBoundarySum(
+    const std::shared_ptr<Mesh>& mesh, const int boundary,
+    std::vector<double>& field ) const
+{
+    // Update left and right boundaries to be their sum.
+    if ( boundary % 2 == 0 )
+    {
+        // Get the boundary nodes.
+        const auto& nodes_left = mesh->getBoundaryNodes( boundary );
+        const auto& nodes_right = mesh->getBoundaryNodes( boundary+1 );
+
+	assert( nodes_left.size() == nodes_right.size() );
+
+        // Apply the boundary condition. Periodic means that the boundaries
+	// are the sum of the left and right side. 
+	for ( std::size_t i = 0; i < nodes_right.size(); ++i )
+	{
+            int right_id = nodes_right[i];
+	    int left_id = nodes_left[i];
+	    double val_left = field[left_id];
+	    double val_right = field[right_id];
+	    field[left_id] += val_right;
+	    field[right_id] += val_left;
+	}
+    }
+}
+
+//---------------------------------------------------------------------------//
+// Complete Boundary sum
+// (-x = 0, +x = 1, -y = 2, +y = 3, -z = 4, +z = 5)
+void PeriodicBoundaryCondition::completeBoundarySum(
+    const std::shared_ptr<Mesh>& mesh, const int boundary,
+    std::vector<std::array<double,3> >& field ) const
+{
+    // Determine if the boundary is on x, y, or z.
+    //int dim = std::floor( boundary / 2 );
+
+    // Update left and right boundaries to be their sum.
+    if ( boundary % 2 == 0 )
+    {
+        // Get the boundary nodes.
+        const auto& nodes_left = mesh->getBoundaryNodes( boundary );
+        const auto& nodes_right = mesh->getBoundaryNodes( boundary+1 );
+
+	assert( nodes_left.size() == nodes_right.size() );
+
+        // Apply the boundary condition. Periodic means that the boundaries
+	// are the sum of the left and right side. 
+	for ( std::size_t i = 0; i < nodes_right.size(); ++i )
+	{
+            for ( int dim = 0; dim < 3; ++dim )
+	    {
+                int right = nodes_right[i];
```

Comment:
you can pull `left` and `right` one level up in the loop to evaluate only once instead of 3 times